### PR TITLE
Convert MemoryMarshal.GetArrayDataReference to a JIT intrinsic

### PR DESCRIPTION
**Pull request by MichalPetryka**
Converts MemoryMarshal.GetArrayDataReference to
an always expand JIT intrinsic and removes the VM intrinsics.

Introduces JIT tests validating the correct behaviour.

Fixes invalid codegen samples from:
https://github.com/dotnet/runtime/issues/58312#issuecomment-993491291

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -3879,6 +3879,21 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
...
+                assert(sig->numArgs == 1);
+
+                GenTree*             array    = impPopStack().val;
+                CORINFO_CLASS_HANDLE elemHnd  = sig->sigInst.classInst[0];
+                CorInfoType          jitType  = info.compCompHnd->asCorInfoType(elemHnd);
+                var_type             elemType = JITtype2varType(jitType);
+
+                GenTree*          index     = gtNewIconNode(0, TYP_I_IMPL);
+                GenTreeIndexAddr* indexAddr = gtNewArrayIndexAddr(array, index, elemType, elemHnd);
+                indexAddr->gtFlags &= ~GTF_INX_RNGCHK;
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:3893]: um.. did you mean `retNode = indexAddr` ?

Comment (MichalPetryka) [reply src/coreclr/jit/importer.cpp:3893]: Thanks, forgot to put that here.

File: src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    if (MemoryMarshal.GetArrayDataReference(a) == 1)
+                    {
+                        return -1;
+                    }
+                }
+
+                return 0;
+            }
+
+            Equals(Problem2(new byte[] { 1 }), 0);
```
Comment (jkotas) [inline src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:82]: I see the Problem1 and Problem2 tests failing with:
```
/__w/1/s/src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:L65 test failed (expected: equal, actual: 2-2).
/__w/1/s/src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:L82 test failed (expected: equal, actual: 0-0).
```
both before and after the change that suggest the problem is somewhere else.

Does the aliasing problem that was mentioned in the original comment actually repro in observable way on current main?

File: src/coreclr/vm/jitinterface.cpp [hunk 3]
```diff
@@ -6975,39 +6975,6 @@ bool getILIntrinsicImplementationForUnsafe(MethodDesc * ftn,
     return false;
 }
 
-bool getILIntrinsicImplementationForMemoryMarshal(MethodDesc * ftn,
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6978]: You can also delete https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/TypeSystem/IL/Stubs/MemoryMarshalIntrinsics.cs

File: src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs (hunk 2 above)
Comment (MichalPetryka) [reply src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:82]: The tests are missing an ! on Equals there, as you can see the actual values are the same here. I'm going to fix it tomorrow together with the missing NREs and invalid IndexOutOfRanges.

As far as the tests here, @SingleAccretion told me on discord that those should still happen on main, haven't verified it though.

**Comment by lambdageek** [comment]
@vargaz After we switched the intrinsic to `OP_CHECK_THIS` in https://github.com/dotnet/runtime/pull/72897, the LLVM AOT build here is failing

```
  /__w/1/s/artifacts/bin/mono/Linux.arm64.Release/cross/linux-arm64/opt: mono_aot_a0U4eG/temp.bc: error: Invalid cast (Producer: 'LLVM11.1.0' Reader: 'LLVM 11.1.0')
  AOT of image /__w/1/s/artifacts/tests/coreclr/Linux.arm64.Release/JIT/Intrinsics/MemoryMarshalGetArrayDataReference_r/MemoryMarshalGetArrayDataReference_r.dll failed.
```

This looks reasonable, doesn't it?
```c
			LLVMBuildLoad2 (builder, IntPtrType (), convert (ctx, lhs, pointer_type (IntPtrType ())), "");
```

**Comment by JulieLeeMSFT** [comment]
@MichalPetryka, I marked this as .NET 8. If this should be .NET 7, please let me know.

**Comment by MichalPetryka** [comment]
.Net 8.0 is the correct target, the PR results in a fair bit of regressions, haven't had time to sit down and figure them out yet.

**Comment by tannergooding** [comment]
@jkotas or @stephentoub, could you reopen this (its locked and needs a maintainer to do so).

We've got known aliasing bugs around using GetArrayDataReference and so even if this causes some small regressions, it resolves real world bugs caused by us relying on the UB of aliasing reference types.

Ideally we'd take this as is now and then look at improving the codegen and bringing us back to the current numbers before .NET 8 ships.

**Comment by MichalPetryka** [comment]
Created a new PR (#79760) since apparently it's not possible to reopen a PR after force pushing.

**Pull request closed without merging**